### ISSUE TITLE
zstream not in use when calling png_write_png second time

### ISSUE THREAD

**Issue by MartinPerry** [glennrp/libpng#301]
I have this code:
```

std::vector<uint8_t> out;
out.reserve(w * h * 4);
png_set_write_fn(png_ptr, &out, PngWriteCallback, NULL);

png_set_compression_strategy(png_ptr, Z_HUFFMAN_ONLY);
png_write_png(png_ptr, info_ptr, PNG_TRANSFORM_IDENTITY, NULL);

std::vector<uint8_t> out1 = std::move(out);
out.clear();
out.reserve(w * h * 4);

png_set_compression_strategy(png_ptr, Z_FILTERED);
png_write_png(png_ptr, info_ptr, PNG_TRANSFORM_IDENTITY, NULL);
```

However, when I call png_write_png second time, program crashes with:
```
libpng warning: zstream not in use (internal error)
libpng error: stream error
```

This happens on Linux, on Windows 10 all is working correctly.

**Comment by benkasminbullock** [comment]
I'm seeing something similar with a recycled PNG file here:

http://matrix.cpantesters.org/?dist=Image-PNG-Libpng%200.48_04

The culprit is this test where $dest_png is repeatedly rewritten:

https://metacpan.org/source/BKB/Image-PNG-Libpng-0.48_04/t%2Fcompress-level.t#L84

However, inspecting the results, the failing Linux machines seem to be using libpng versions 1.5.13:

http://www.cpantesters.org/cpan/report/10849ee4-3980-11eb-88e7-3d1b1f24ea8f
http://www.cpantesters.org/cpan/report/09b2a5c0-3980-11eb-88e7-3d1b1f24ea8f

I see successful outcomes with libpng versions from 1.6:

http://www.cpantesters.org/cpan/report/d9b93528-396b-11eb-bb18-7c761f24ea8f
http://www.cpantesters.org/cpan/report/564ffaea-3930-11eb-b23a-96691f24ea8f
http://www.cpantesters.org/cpan/report/def6c368-3982-11eb-84bc-edd243e66a77

**Comment by jbowler** [comment]
Yes, **writing a PNG with the same `png_struct`/`png_info` more than once is not supported**.  

I believe the original intent was that `png_info` could be reused but it wasn't explicit in the design and, while it might have worked in libpng-12 it was broken in, I think, libpng-14 (or maybe libpng-15).

To see the most that you can do in this scenario take a look at `contrib/tools/pngcp.c` and, in particular, the function `write_png`.  This does a lot more than your code but it's pretty much the same in structure and intent (it's copying a PNG from png_read_png).  Notice that it always creates a new `png_struct`.  `png_struct` should not contain anything about the PNG to be written; that should all come from `png_info` (that is the original design).

Higher up in the file the function `text_stash` explains in detail what the problem is with reuse of `png_info`.  If you don't write text chunks you should be able to reuse the `png_info` but if you want to be safe don't - it was broken once it can be broken again and it isn't being tested in any version of libpng!  If you are doing what pngcp is doing - making a complete copy of an external PNG, you need to reuse the `png_info` so you need to do the `text_stash` stuff.

Don't assume that this will be fixed in future versions; libpng-17 has a fix but libpng-17 has never been released and it is a significant rewrite of libpng (it's more like libpng-20).  It seems unlikely that the next version of libpng, maybe libpng-18, will have such extensive changes as minimal resource is available for development.
